### PR TITLE
[SPARK-47248][SQL][COLLATION] Improved string function support: contains

### PR DESCRIPTION
**Pull request by uros-db**
### What changes were proposed in this pull request?
Extend built-in string functions to support non-binary, non-lowercase collation for: contains.

### Why are the changes needed?
Update collation support for built-in string functions in Spark.

### Does this PR introduce _any_ user-facing change?
Yes, users should now be able to use COLLATE within arguments for built-in string function CONTAINS in Spark SQL queries, using non-binary collations such as UNICODE_CI.

### How was this patch tested?
Unit tests for queries using "collate" (CollationSuite).

### Was this patch authored or co-authored using generative AI tooling?
Yes.

### CODE REVIEW

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 1]
```diff
@@ -343,19 +346,33 @@ public boolean contains(final UTF8String substring) {
...
+    if (substring.numBytes == 0) {
+      return true;
+    } else if (this.numBytes == 0) {
+      return false;
+    }
+    String pattern = substring.toString();
+    CharacterIterator target = new StringCharacterIterator(this.toString());
+    Collator collator = CollationFactory.fetchCollation(collationId).collator;
+    StringSearch stringSearch = new StringSearch(pattern, target, (RuleBasedCollator) collator);
+    stringSearch.reset();
```
Comment (cloud-fan) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:369]: We need to call `reset` on a fresh instance?

**Review by cloud-fan — APPROVED** [review]

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 1 above)
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: indeed, removed

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 2]
```diff
@@ -343,19 +346,33 @@ public boolean contains(final UTF8String substring) {
...
+  private boolean collatedStringSearch(final UTF8String substring, int collationId) {
+    if (substring.numBytes == 0) {
+      return true;
+    } else if (this.numBytes == 0) {
+      return false;
+    }
+    String pattern = substring.toString();
+    CharacterIterator target = new StringCharacterIterator(this.toString());
+    Collator collator = CollationFactory.fetchCollation(collationId).collator;
+    StringSearch stringSearch = new StringSearch(pattern, target, (RuleBasedCollator) collator);
```
[the remark this answers is not in the document]
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: done!

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 3]
```diff
@@ -145,6 +149,18 @@ public Collation(
     }
   }
 
+  /**
+   * Auxiliary methods for collation aware string operations.
+   */
+
+  public static StringSearch getStringSearch(final UTF8String left, final UTF8String right,
+                                          final int collationId) {
```
Comment (cloud-fan) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:157]:
```suggestion
  public static StringSearch getStringSearch(
      final UTF8String left,
      final UTF8String right,
      final int collationId) {
```

Comment (cloud-fan) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:157]: are we going to use it for startsWith and others?

Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:157]: definitely, that's in ([SPARK-47295](https://issues.apache.org/jira/browse/SPARK-47295)) and will serve as an onboarding task for a new eng (of course, I will help them out with this)

**Comment by cloud-fan** [comment]
The GA jobs all passed: https://github.com/uros-db/spark/actions/runs/8186876833/job/22395549669

merging to master, thanks!

**Pull request closed without merging**